### PR TITLE
Image loop improvement

### PR DESCRIPTION
**Pull request by InformatiqueProg**
Add width/height for image and original image

### CODE REVIEW

**Comment by gillesbourgeat** [comment]
Hello,

You break four unit tests
https://travis-ci.org/thelia/thelia/builds/87851013

Best regards

**Comment by gillesbourgeat** [comment]
It is not possible to apply your pull request in Thelia.
This change makes the image loop too greedy in disk resource.

thank you for your proposition.

**Comment by InformatiqueProg** [comment]
And if i add an argument  'with_image_size' ?

**Comment by gillesbourgeat** [comment]
No,
It is possible for you, to delegate this behavior in a module.

```php
class MyLoopImage extends Image
{
    /**
     * Use this method in order to add fields in sub-classes
     * @param LoopResultRow $loopResultRow
     * @param object|array $item
     */
    protected function addOutputFields(LoopResultRow $loopResultRow, $item)
    {
        // your code
    }
}
```

**Comment by lunika** [comment]
> This change makes the image loop too greedy in disk resource.

Can you explain why @gillesbourgeat  ?

**Comment by InformatiqueProg** [comment]
Okay,
i will do it in a module.
I see some post who says that getimagesize load entire image, that's why ?

**Comment by roadster31** [comment]
getimagesize() is well known to be slow on remote files, but as far as I know, there's no specific performance issues on local files. Thus, adding another getimagesize() will have little impact on the loop performance.

Let's make some tests :)

**Comment by roadster31** [comment]
Here are the results for 50 iterations of the following code on a live server, with and without getimagesize(), for 1572 images :

```
{loop type="product" name="testprod"}
    {loop type="image" name="test" product=$ID}
    {/loop}
{/loop}
```
The results are showing an average cost of 0.46 seconds when the getimagesize() version is used. A very small difference.

Giving this results, I'm supporting this PR, as getting image sizes may be useful for the front designer.

img-2015-10-29 10 49 10

File: core/lib/Thelia/Core/Template/Loop/Image.php [hunk 1]
```diff
@@ -361,11 +361,19 @@ public function parseResults(LoopResult $loopResult)
                 // Dispatch image processing event
                 $this->dispatcher->dispatch(TheliaEvents::IMAGE_PROCESS, $event);
 
+                $originalImageSize = getimagesize($sourceFilePath);
```
Comment (bibich) [inline core/lib/Thelia/Core/Template/Loop/Image.php]: I think you should use `$event->getOriginalFileUrl()` instead of `$sourceFilePath` here.

**Comment by InformatiqueProg** [comment]
Okay,
i'll change this, and test it.
While i am, what about add an argument 'with_image_size' to prevent this ?

File: core/lib/Thelia/Core/Template/Loop/Image.php (hunk 1 above)
Comment (InformatiqueProg) [inline core/lib/Thelia/Core/Template/Loop/Image.php]: You're sure ?
 $event->getOriginalFileUrl() return an url => http://mydomains.fr/cache/images/product/on_air_2-6.jpg
it seem that remote file are not the best choice for getImageSize()...

**Comment by Soldras** [comment]
Hello,
the IMAGE_WIDTH and IMAGE_HEIGHT parameters are useless because if you don't pass width or height, you get the original file size. 

I don't know what you want to do with this parameters but the loop give you ORIGINAL_IMAGE_URL and IMAGE_URL. In front office if you need image size, it's the parameters you have pass to the loop.
If you need orignal size to perform some front processing, you can do it with some js.

It's not a need for Thelia so it's better to do a module.

**Comment by roadster31** [comment]
> the IMAGE_WIDTH and IMAGE_HEIGHT parameters are useless because if you don't pass width or height, you get the original file size. 
No. You can give only width or only height.

Additionally, without IMAGE_WIDTH and IMAGE_HEIGHT, you have to *copy* the numeric width and height passed to the loop, which may cause errors / inconsistencies.

> If you need orignal size to perform some front processing, you can do it with some js.
It's very easy to provide the dimension parameters from the loop. Why bother the front developer with complicated stuff to get them ?

**Comment by lunika** [comment]
:+1: with @roadster31 :-)

**Comment by Soldras** [comment]
> No. You can give only width or only height.
It's true but if you have one size element you can calculate the other.

> It's very easy to provide the dimension parameters from the loop. Why bother the front developer with complicated stuff to get them ?
It's better to let front dev to front developers. Image size is like the spoon, it doesn't exit in front dev (Cf Matrix ).
If you have a basic image you rules it by css. If you want some dynamic view you do it in js.

**Comment by roadster31** [comment]
Once again, providing these dimensions have not any performance impact, and may help some front developers. So let's provide them !

I'll not argue any longer. I suggest a vote to decide if we'll accept this PR.

**Comment by Soldras** [comment]
The real question is : 
it's interesting to loose 6% answer time to 0.1% use case ?

**Comment by lunika** [comment]
> The real question is :
it's interesting to loose 6% answer time to 0.1% use case ?

This is why loop were designed for. Be the more generic possible and provide much information as possible.

**Comment by roadster31** [comment]
6% ? The average impact is 0.46 second for 1572 image, which gives an average impact of 0.00003 second per image.

**Comment by Soldras** [comment]
> This is why loops were designed for. Be the more generic possible and provide much information as possible.

Being generic is important. But being generic not involve being slow because we want to cover all particular use case. This type of use case must be cover by module.

> 6% ? The average impact is 0.46 second for 1572 image, which gives an average impact of 0.00003 second per image. No big deal !

8.057 x 100 / 8,5156 = 94,6145... So theres is 5,4% performance lost. It's valid for each image.

**Comment by roadster31** [comment]
> 8.057 x 100 / 8,5156 = 94,6145... So theres is 5,4% performance lost. It's valid for each image.

True. 5,4% for a single image. This means 0.00003 seconds. Is it a performance problem ?

**Comment by roadster31** [comment]
Well, stop feeding the troll.

Let's vote.

**Comment by lunika** [comment]
Let's vote : 

OpenStudio : :-1: 
Community : :+1: 

Hoops, stop feeding the troll :-)

**Comment by bilhackmac** [comment]
It can be a performance problem if images are hosted on a CDN (incoming feature)

**Comment by Alban-io** [comment]
@bilhackmac :+1:

**Comment by InformatiqueProg** [comment]
Nobody want's my argument 'with_image_size' ? :disappointed_relieved:
It will only use on somes loop...

**Comment by Soldras** [comment]
@InformatiqueProg :
Dalek

**Comment by lunika** [comment]
@Soldras he already explain twice, please read the discussion !

This kind of argument already exist and is used in the `Product` loop for example : https://github.com/thelia/thelia/blob/master/core/lib/Thelia/Core/Template/Loop/Product.php#L357

**Comment by Soldras** [comment]
@lunika : he didn't explain why he need these outputs

**Comment by InformatiqueProg** [comment]
Like @roadster31 said, 
We need width and height for front end integration of some image (JS gallery, etc.)

**Comment by lunika** [comment]
I apologize for my troll...

Have fun guys

**Pull request closed without merging**